src/main/kotlin/snyk/code/annotator/SnykCodeAnnotator.kt

```diff
@@ -0,0 +1,61 @@
+package snyk.code.annotator
+
+import ai.deepcode.javaclient.core.MyTextRange
+import ai.deepcode.javaclient.core.SuggestionForFile
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.ExternalAnnotator
+import com.intellij.lang.annotation.HighlightSeverity
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiFile
+import io.snyk.plugin.snykcode.core.AnalysisData
+
+class SnykCodeAnnotator : ExternalAnnotator<PsiFile, Unit>() {
+    // overrides needed for the Annotator to invoke apply(). We don't do anything here
+    override fun collectInformation(file: PsiFile): PsiFile = file
+    override fun doAnnotate(collectedInfo: PsiFile?): Unit = Unit
+
+    override fun apply(psiFile: PsiFile, annotationResult: Unit, holder: AnnotationHolder) {
+        val suggestions = getIssuesForFile(psiFile)
+
+        suggestions.forEach { suggestionForFile ->
+            val severity = severity(suggestionForFile)
+            suggestionForFile.ranges.forEach {
+                val textRange = textRange(psiFile, it)
+                if (!textRange.isEmpty) {
+                    val annotationBuilder =
+                        holder.newAnnotation(severity, annotationMessage(suggestionForFile)).range(textRange)
+                    annotationBuilder.create()
+                }
+            }
+        }
+    }
+
+    fun severity(suggestion: SuggestionForFile): HighlightSeverity {
+        return when (suggestion.severity) {
+            3 -> HighlightSeverity.ERROR
+            2 -> HighlightSeverity.WARNING
+            1 -> HighlightSeverity.WEAK_WARNING
+            else -> HighlightSeverity.INFORMATION
+        }
+    }
+
+    private fun getIssuesForFile(psiFile: PsiFile): List<SuggestionForFile> {
+        ProgressManager.checkCanceled()
+        return AnalysisData.instance.getAnalysis(psiFile)
+    }
+
+    fun annotationMessage(suggestion: SuggestionForFile): String {
+        return buildString {
+            val title = if (suggestion.title.isNotBlank()) " " + suggestion.title + "." else ""
+            append("Snyk Code:$title ${suggestion.message}")
+        }
+    }
+
+    fun textRange(psiFile: PsiFile, snykCodeRange: MyTextRange): TextRange {
+        val document = psiFile.viewProvider.document ?: return TextRange.EMPTY_RANGE
+        val lineOffSet = document.getLineStartOffset(snykCodeRange.startRow - 1) + snykCodeRange.startCol
```

Comment:
Don't know if it's 0 or 1-base indexing here - just verifying if that could be prone to IndexOutOfBoundsException?

Reply:
`if (line < 0 || document.lineCount <= line) return TextRange.EMPTY_RANGE`